### DISCUSSION TITLE
UEFI LTSP clients only boot with http when dnsmasq runs in dhcp-proxy mode

### DISCUSSION THREAD

**Discussion by X-Dimension-88MPH** [ltsp/ltsp#922] [Q&A]
We are using http instead of tftp for faster boot times with old BIOS based computers without a problem.
But now we have some new UEFI based computers that are always booting with tftp instead of http and so the boot times are much slower than with the old clients and http. 

After some testing we found out that the UEFI clients also using http on boot when we are set up dnsmasq in dhcp proxy mode behind a isc-dhcp-server instead of using dnsmasq as a standalone dhcp-server.

We don't understand why UEFI clients are booting fine with http when dnsmasq is running as dhcp-proxy and are using tftp when dnsmasq is running as standalone server.

Our dnsmasq configuration looks like this: 

**1. Dnsmasq in Proxy Mode**

```
# Configure dnsmasq for LTSP
# Documentation=man:ltsp-dnsmasq(8)

# For additional local dnsmasq configuration like DNS blacklisting, it's
# recommended to use separate /etc/dnsmasq.d/local.conf files,
# so that they're not lost if you ever (re)run `ltsp dnsmasq`.

# port=0 disables the DNS service of dnsmasq
port=0

# enable-tftp enables the TFTP service of dnsmasq
enable-tftp

# FHS 2.3+ recommends /srv for tftp (debian #477109, LP #84615)
tftp-root=/srv/tftp

# If another DHCP server is present on the network, a proxy range may be used
# instead. This makes dnsmasq provide boot information but not IP leases.
dhcp-range=set:proxy,192.168.18.0,proxy,255.255.255.0

# Specify the DNS server. 0.0.0.0 means the machine running dnsmasq.
# DNS_SERVER in ltsp.conf is preferred as it reaches proxy DHCP clients.
dhcp-option=option:dns-server,192.168.18.201

# Set some tags to be able to separate client settings later on.
dhcp-match=set:X86PC,option:client-arch,0
dhcp-match=set:X86-64_EFI,option:client-arch,7
# Due to rfc4578 errata, sometimes BC_EFI=9 is misused instead of X86-64_EFI=7:
dhcp-match=set:X86-64_EFI,option:client-arch,9

# Check if iPXE has the features we need: http://ipxe.org/howto/dhcpd
# http && menu && ((pxe && bzimage) || efi)
dhcp-match=set:ipxe-http,175,19
dhcp-match=set:ipxe-menu,175,39
dhcp-match=set:ipxe-pxe,175,33
dhcp-match=set:ipxe-bzimage,175,24
dhcp-match=set:ipxe-efi,175,36
tag-if=set:ipxe,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-pxe,tag:ipxe-bzimage
tag-if=set:ipxe,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-efi

pxe-service=tag:proxy,tag:ipxe,X86PC,"ltsp.ipxe","http://${proxydhcp/dhcp-server}/ltsp/ltsp.ipxe"
pxe-service=tag:proxy,tag:ipxe,X86-64_EFI,"ltsp.ipxe","http://${proxydhcp/dhcp-server}/ltsp/ltsp.ipxe"

# Specify the boot filename for each tag, relative to tftp-root.
# If multiple lines with tags match, the last one is used.
# See: https://www.syslinux.org/wiki/index.php?title=PXELINUX#UEFI
dhcp-boot=tag:!ipxe,tag:X86PC,ltsp/undionly.kpxe
dhcp-boot=tag:!ipxe,tag:X86-64_EFI,ltsp/snponly.efi
dhcp-boot=tag:ipxe,ltsp/ltsp.ipxe

``` 
On our isc-dhcp-server (192.168.18.201) we are only adding the line "next-server 192.168.18.40;" but everything else is the default Ubuntu configuration. (192.168.18.40 is the LTSP/dnsmasq server in this case)

**2. Dnsmasq in Standalone Mode**

```
# Configure dnsmasq for LTSP
# Documentation=man:ltsp-dnsmasq(8)

# For additional local dnsmasq configuration like DNS blacklisting, it's
# recommended to use separate /etc/dnsmasq.d/local.conf files,
# so that they're not lost if you ever (re)run `ltsp dnsmasq`.

# port=0 disables the DNS service of dnsmasq
port=0

# enable-tftp enables the TFTP service of dnsmasq
enable-tftp

# FHS 2.3+ recommends /srv for tftp (debian #477109, LP #84615)
tftp-root=/srv/tftp

# Log lots of extra information about DHCP transactions
#log-dhcp

# IP ranges to hand out, usually on the internal LTSP subnet of 2-NIC setups
dhcp-range=192.168.46.11,192.168.46.190,12h

# Specify the DNS server. 0.0.0.0 means the machine running dnsmasq.
# DNS_SERVER in ltsp.conf is preferred as it reaches proxy DHCP clients.
dhcp-option=option:dns-server,192.168.46.10

#Default Gateway
dhcp-option=3,192.168.46.10

# Set some tags to be able to separate client settings later on.
dhcp-match=set:X86PC,option:client-arch,0
dhcp-match=set:X86-64_EFI,option:client-arch,7
# Due to rfc4578 errata, sometimes BC_EFI=9 is misused instead of X86-64_EFI=7:
dhcp-match=set:X86-64_EFI,option:client-arch,9

# Check if iPXE has the features we need: http://ipxe.org/howto/dhcpd
# http && menu && ((pxe && bzimage) || efi)
dhcp-match=set:ipxe-http,175,19
dhcp-match=set:ipxe-menu,175,39
dhcp-match=set:ipxe-pxe,175,33
dhcp-match=set:ipxe-bzimage,175,24
dhcp-match=set:ipxe-efi,175,36
tag-if=set:ipxe,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-pxe,tag:ipxe-bzimage
tag-if=set:ipxe,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-efi

pxe-service=tag:ipxe,X86PC,"ltsp.ipxe","http://${dhcp-server}/ltsp/ltsp.ipxe"
pxe-service=tag:ipxe,X86-64_EFI,"ltsp.ipxe","http://${dhcp-server}/ltsp/ltsp.ipxe"

# Specify the boot filename for each tag, relative to tftp-root.
# If multiple lines with tags match, the last one is used.
# See: https://www.syslinux.org/wiki/index.php?title=PXELINUX#UEFI
dhcp-boot=tag:!ipxe,tag:X86PC,ltsp/undionly.kpxe
dhcp-boot=tag:!ipxe,tag:X86-64_EFI,ltsp/snponly.efi
dhcp-boot=tag:ipxe,ltsp/ltsp.ipxe

``` 
I hope somebody could find the configuration error in the standalone setup.

Thanks for help

**Comment by X-Dimension-88MPH** [comment]
After some more tests it looks like the dhcp-proxy mode is not the reason why it works only on one of our setups!

I've installed an ISC-DHCP-Server on a test-network and using a copy of our working LTSP-Server behind it and the UEFI clients also using tftp instead of http now. :(

So the last difference between the working setup and all others is, that we are using a Samba active-directory domain controller with Bind DNS server there and not only a ISC-DHCP-Server or a standalone dnsmasq as dhcp-server.

But what is the difference between a samba active-directory domain controller and a plain isc-dhcp-server when it comes to PXE boot? Is PXE boot depending on DNS when using http? 
DNS is the only difference i see between the working setup and all others.

**Reply by X-Dimension-88MPH** [reply]
Using isc-dhcp-server as dhcp-server and dnsmasq as proxy also works now with UEFI clients and HTTP.
I don't really know why it doesn't work before, maybe i had a typo somewhere.

But with dnsmasq as standalone dhcp server UEFI clients still switch to tftp instead of http.

I can live with the isc-dhcp-server/dnsmasq setup for now, but if anybody has an idea how to get it work with dnsmasq only, i'll be thankful for any solution. :)

**Reply by techieshark** [reply]
Very late to the party here, but in case useful to you or anyone else, there are some 'tcpdump' commands that might be useful to run to observe the actual DHCP traffic and see if that gives any hints. For more on that, see https://github.com/ltsp/ltsp/discussions/957
